### DISCUSSION TITLE
How to avoid textbox scaling as text inside is added.

### DISCUSSION THREAD

**Discussion by soham-bhosale** [fabricjs/fabric.js#9833] [Q&A]
I am using fabric js 5.3.0, and I have a use case where in user will add text into a text area and then it is displayed on canvas.

So the line in text area (input field) should respond to line in canvas. As text increases, font size should decrease till some threshold and then stop.
Same is with height.
There are also some other things like font change, bold, italic, underline.

Basically, the dimensions of textbox should remain same.

No matter what i try, i am not being able to make this work.
I tried to utilize the textlines number to keep track and decrease the font size when textlines increase but that is not working very well for very long text (50+ chars.)

**Comment by asturur** [comment]
> No matter what i try, i am not being able to make this work.

Where is what you tried? show us.

**Reply by soham-bhosale** [reply]
First I tried to using width. So when the textbox obj is created, I use the width at that and set is as maxWidth.
Then I compare it with the current width and adjust the font size accordingly.
The issue is that even if i change the text back, the font doesn't roll back. So once, I reduce the font, its not increasing if the characters are then reduced.
BTW, I am using staticCanvas and the text is added to a textbox by user.
```
if (currentObject.width > currentObject.fixedWidth) {
    currentObject.fontSize *= currentObject.fixedWidth / (currentObject.width + 1);
    currentObject.width = currentObject.fixedWidth;
}
```

For second, I did splitByGraphMe=true for object and i am checking that if the number of line the text split into has increased. If so, reduce the font. This one works quite good but for only upto 20-25 chars after that it starts to fail (Don't know why). Its a wild approach but it's a honest approach.
Also, there are things like maxLines and maxChars here please ignore them.
```
changeText: function(event) {
                let currentObject = this.canvas.getObjects()[parseInt(event.target.getAttribute('idx'))];
                if (!currentObject) {
                    return;
                }

                let oldText = currentObject.text;
                let currLines = currentObject.textLines;
                let transformedText = event.target.value.split('\n')
                    .slice(0, parseInt(currentObject.customAttributes.maxLines))
                    .map((line) => {
                        return line.substring(0, parseInt(currentObject.customAttributes.maxCharacters ?? 0));
                    })
                    .join('\n');


                currentObject.set('text', transformedText);

                if (!currentObject.customAttributes.maxSize) {
                    currentObject.customAttributes.maxSize = currentObject.fontSize;
                }
                let maxSize = parseInt(currentObject.customAttributes.maxSize);
                if (event.inputType != 'insertLineBreak') {
                    currentObject.set('fontSize', maxSize);
                    while (currentObject.textLines.length > currLines.length) {
                        if (currentObject.customAttributes.minSize && currentObject.fontSize <= currentObject.customAttributes.minSize) {
                            break;
                        }
                        currentObject.set('fontSize', --maxSize);
                    }
                }

                event.target.value = transformedText;
                this.canvas.requestRenderAll();
            }
 ```
 
 Do you have any other solution?

**Reply by asturur** [reply]
please put it in a working environment where it can be tried, jsfiddle or codesandbox

**Reply by soham-bhosale** [reply]
@asturur Sorry for the late reply. I took some time off. Looks like that did the thing and I have figured out my certain use case.
Here is the working [fiddle](https://jsfiddle.net/dextroz0808/0m5xcowz/9/) link.

I have also managed to embed max lines, max characters per line and min font size for textbox.

I think I over-engineered my initial solution trying to use splitByGraphme and stuff like that.

Also, notice how I need to render the canvas after setting font(line 40). Is there any better way? I suppose initDimensions() is working correctly but then the rendering is not done till we scale the font. 

Adding code here just in case:-
**JS**
```
let canvas = new fabric.StaticCanvas('canvas');
var t1 = new fabric.Textbox('My Text', {
    width: 200,
    top: 50,
    left: 50,
    fontSize: 16,
    textAlign: 'center',
    fixedWidth: 200,
});

/* Pre Defined Values */
t1.minFont = 12;
t1.maxChars = 40;
t1.maxLines = 12;


t1.fixedHeight = t1.height*(t1.maxLines ?? 1);
t1.maxFont = t1.fontSize;


canvas.add(t1);

document.getElementById('input').addEventListener('input', function (event){
  resizeTextBox(event,t1);
})

function resizeTextBox(event,t1){

let transformedText = event.target.value.split('\n')
                    .slice(0, parseInt(t1.maxLines ?? 0))
                    .map((line) => {
                        return line.substring(0, parseInt(t1.maxChars ?? 0));
                    })
                    .join('\n');

  t1.text = transformedText;
  event.target.value = transformedText;
	t1.fontSize = t1.maxFont;
  // Need refactoring;
	canvas.renderAll();
//  t1.initDimensions();
	//fabric.util.clearFabricFontCache();
  if (t1.width > t1.fixedWidth) {
  	t1.fontSize = t1.fontSize * (t1.fixedWidth / (t1.width + 1));
  }
  if (t1.height > t1.fixedHeight) {
  	t1.fontSize = t1.fontSize * (t1.fixedHeight/ (t1.height + 1));
  	t1.height = t1.fixedHeight;
  }
  
  t1.width = t1.fixedWidth;
  console.log(t1.width, t1.fontSize, t1.textAlign);
  canvas.renderAll();
  
  if (t1.minFont && t1.fontSize < t1.minFont) {
  	t1.text = t1.text.slice(0,-1);
    event.target.value = t1.text;
    resizeTextBox(event,t1);
  }
}
```
**HTML**
```
<textarea id="input"></textarea>
<canvas id="canvas" width=500 height=500></canvas>
```
**CSS**
```
#canvas {
  border: 1px solid grey;
  display: flex;
}
```
